Stowable trait

Lucretiel opened the pull request:
- Addded unsafe trait Stowable, which asserts that a type has no uninit bytes
- Implemented Stowable for several built-in types- pointers, arrays, primitives
- Removed several TODO comments after validating correct behavior

Fixes #8 

@KrishnaSannasi

Lucretiel committed:
Initial version of a Stowable trait which punts the Uninit soundness issues

- Addded unsafe trait Stowable, which asserts that a type has no uninit bytes
- Implemented Stowable for several built-in types- pointers, arrays, primitives
- Removed several TODO comments after validating correct behavior

RustyYato commented on src/lib.rs:
```diff
@@ -134,13 +137,130 @@
+    /// For objects with padding bytes or other uninitialized bytes that you
+    /// still want to use with stowaway, set this to true. It *is* safe to
+    /// implement `Stowable` for *any* type if MUST_BOX is true.
+    const MUST_BOX: bool = false;
```
I don't see the point in `MUST_BOX`, let's simplify the `unsafe` obligations to "the type must either be larger than a pointer, must be more aligned than a pointer, or must not contain *any* uninit bytes". You can document the case where using a `Stowaway<Box<T>>` won't do an extra allocation, so if a type does contain any uninit bytes it can be boxed at no extra cost.

(I'm trying to make it simpler to reason about `Stowable` so that it is easier to verify the correctness of this crate).

Lucretiel replied:
Oh; for some reason I had assumed you had a need to have `MaybeUninit` types in `Stowaway`. I'm definitely in favor of simplifying.

RustyYato commented on src/lib.rs:
```diff
@@ -134,13 +137,130 @@
+/// a size that is a multiple of their alignment, so there's no type that is
+/// stowable but an array is not stowable.
+// TODO: use rustversion to opt-in to const generics if this is a nightly build
+macro_rules! stowable_array {
```
Maybe a nightly version of this crate can use const-generics to get all array impls? Not necessary, but it may be nicer.

Lucretiel replied:
Yeah, I think we can detect that with https://github.com/dtolnay/rustversion and make it work without a feature flag. I definitely don't want to make the crate nightly-only, so I've left it as a TODO for now.

RustyYato commented on src/lib.rs:
```diff
@@ -134,13 +137,130 @@
+// when calling from_raw.
+unsafe impl<T> Stowable for *const T {}
+unsafe impl<T> Stowable for *mut T {}
+unsafe impl<T> Stowable for &T {}
```
Where's `&mut T`, `Option<&T>`, and `Option<&mut T>`?

Lucretiel replied:
Whoops, good catch

RustyYato commented on src/lib.rs:
```diff
@@ -134,13 +137,130 @@
+unsafe impl Stowable for () {}
+
+/// Vec is always (ptr, size, cap)
+unsafe impl<T> Stowable for Vec<T> {}
```
Also a `Vec<_>` is always bigger than a `usize`, so it can't be stored inline.

Lucretiel replied:
`Stowable` just means that the type can be stored in a `Stowaway`, not that it specifically will be packed instead of boxed.

RustyYato replied:
Yes, Just a note about the documentation :)

RustyYato commented on src/lib.rs:
```diff
@@ -134,13 +137,130 @@
+unsafe impl<T> Stowable for Vec<T> {}
+
+/// MaybeUninit can obviously contain uninit bytes
+unsafe impl<T> Stowable for MaybeUninit<T> {
```
As per the first comment, `MaybeUninit<T>` shouldn't implement `Stowable`, if users want to use it, then they must first allocate their type using one of `Box<_>`, `Rc<_>`, or `Arc<_>`. This can be documented.

RustyYato commented on src/lib.rs:
```diff
@@ -173,28 +293,17 @@ use SizeClass::*;
-pub struct Stowaway<T> {
-    // TODO: Reimplemnt this as union, once we can have non-copy fields in a
+pub struct Stowaway<T: Stowable> {
+    // TODO: Reimplement this as union, once we can have non-copy fields in a
```
You can't use a union here, I assume you want
```
union Stowable<T: Stowable> {
    ptr: *mut T,
    value: T,
}
```
But this would make `Stowable` the same size as `T` when `sizeof(T) > sizeof(ptr)`, which is not what we want.

Lucretiel replied:
There's an earlier draft of this project- I'm not sure if it's in the git history- where this does actually work, using a trait and an associated type, though I ended up getting rid of it because of the non-`Copy` restriction I mentioned.

RustyYato replied:
Oh, I see. I think that would require const generics because you need to choose the layout based on the size and alignment, but that's a nice idea!

Lucretiel committed:
rustdoc updates, added f32/f64

RustyYato commented on src/lib.rs:
```diff
@@ -173,28 +293,17 @@ use SizeClass::*;
-        if mem::align_of::<T>() > mem::align_of::<*mut T>() {
-            // Need to unconditionally box values that don't match our
-            // alignment, even if sizeof(T) <= sizeof(Self)
+        if T::MUST_BOX {
```
As per the first comment, this code get's removed

RustyYato commented on src/lib.rs:
```diff
@@ -243,10 +352,11 @@ impl<T> Stowaway<T> {
+                        // are not always copied this means that MIRI will see zeros where
+                        // the paddings bytes are suppoed to be and allows it. But this is
+                        // *still* undefined behavior! So in order to make it easier for MIRI
+                        // to catch this behavior, we manually copy padding bytes
```
Can you link to the MIRI issue here?

RustyYato commented on src/lib.rs:
```diff
@@ -134,13 +137,130 @@
+    Rc<T> Option<Rc<T>>
+    RcWeak<T> Option<RcWeak<T>>
+    NonNull<T> Option<NonNull<T>>
+}
```
Why isn't `Arc` here? Also you can use a feature flag `alloc` to enable the `alloc` crate. This will allow you to use this crate on `no_std`.

The way I like to handle this is by doing:

```rust
#[cfg(all(not(feature = "std"), feature = "alloc"))]
extern crate alloc as std;
```
Then change all imports to use `core::` instead of `std::` except where `std` in necessary (for allocations or for `std` specific stuff)
You can also then gate all allocation related stuff on 
```rust
#[cfg(any(feature = "std", feature = "alloc"))]
```
But maybe for another PR.

Lucretiel replied:
Oh, for some reason I thought that `Arc` was std-only, because of atomic operations.

All the stuff you mentioned about an std feature is in this crate already, though I can get rid of it since right now `Arc` is the only thing it was being used for.

Lucretiel replied:
Also, we can't put `alloc` behind a feature flag, because the standard operation of the library includes boxing larger types. We have to use it unconditionally.

RustyYato reviewed (changes requested):
I don't think the `Stowable` trait should need `MUST_BOX`, it just complicates things for no real gain.

RustyYato commented on src/lib.rs:
```diff
@@ -275,6 +295,9 @@ use SizeClass::*;
 ///
+/// Types stored in a `Stowaway` must implement [`Stowable`] in order to
+/// prevent unsound behavior related to uninitialized bytes.
+///
```
Link to #8 here

Lucretiel commented:
Thanks for the comments; I'm happy to get rid of `MUST_BOX`. Will fix these tonight.

Lucretiel committed:
Code review changes per #12

- Removed MUST_BOX for simplicity
- Removed the "std" feature, as Arcs are available in "alloc"
- Added links to the relevant github issues in the soundness docs

Lucretiel commented:
Edits completed

Lucretiel committed:
Removed failing test; fixed missing newline

RustyYato commented on src/lib.rs:
```diff
@@ -134,13 +137,129 @@
+unsafe impl<T> Stowable for &T {}
+unsafe impl<T> Stowable for Option<&mut T> {}
+unsafe impl<T> Stowable for Option<&T> {}
+unsafe impl<T> Stowable for &mut T {}
```
I missed this last time, these could take `T: ?Sized`

RustyYato commented on src/lib.rs:
```diff
@@ -134,13 +137,129 @@
+///
+/// Structs which are `repr(transparent)` are `Stowable` if their inner field
+/// is `Stowable`.
+pub unsafe trait Stowable: Sized {}
```
I don't think we need to enforce `Sized` here, it prevents other people from using `Stowable` as a super-trait if they want dynamic dispatch, and `Stowaway` already requires `T: Sized` by default.

RustyYato commented on src/lib.rs:
```diff
@@ -134,13 +137,129 @@
+/// is `Stowable`.
+pub unsafe trait Stowable: Sized {}
+
+// TODO: determine when enums are safe to stow. For instance, is an
+// Option<usize> safe to stow? if the representation is IIIIIIII?UUUUUUU,
+// where I is the int part and ? is the discriminant, probably not.
+// At the very least, types where sizeof(Option<T>) == sizeof(T) are always
+// safe to stow as long as T is safe to stow.
```
`Option`'s representation is unspecified except for non-null optimization, so we can't say anything about it beyond that.

Lucretiel committed:
Updated Sized bounds

Lucretiel commented:
Sized changes added

KrishnaSannasi commented:
LGTM

RustyYato commented on src/lib.rs:
```diff
@@ -173,36 +297,27 @@ use SizeClass::*;
+        // Previously we checked the alignment of T, but the only way for
+        // alignof(T) > alignof(ptr) to be true is if
+        // sizeof(t) == 0 or sizeof(T) > sizeof(ptr), both of which are
+        // handled.
```
Is this documented somewhere? If not can you add this check because there shouldn't be any cost to doing so, const propogation should eliminate the check.

Lucretiel replied:
Yup:
https://stackoverflow.com/questions/61256701/is-it-possible-to-have-a-type-with-a-larger-alignment-than-its-own-size
https://doc.rust-lang.org/reference/type-layout.html#size-and-alignment
> The size of a value is always a multiple of its alignment.

Lucretiel replied:
I'm happy to restore it, but we require this property to be true in order for `[T; N]` to be stowable when `T: Stowable`, so I was happy to assume it here.

RustyYato replied:
Ok, that works for me

RustyYato commented:
This looks good to me
